### PR TITLE
[Core]Use worker shim PID to check worker registration

### PR DESCRIPTION
**Pull request by chenk008**
## Why are these changes needed?

Currently raylet will check whether the ray worker process PID is the same with forked process PID.  After setuping the runtime env, the shim process will use `exec` to start worker process, so the shim process PID is the same with worker process. 

And we want use runtime shim process to start this container, it will fork several times to start the worker process, the worker process PID will be different from shim process PID, so the work will failed to register. Add the worker shim PID in the registration, and use this PID to check whether the worker is legal.

## Related issue number

https://github.com/ray-project/ray/pull/16370
https://github.com/ray-project/ray/pull/15976

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by chenk008** [comment]
The CI `doc:doc_code_runtime_env_example` always failed , because it will install ray[serve]  release version. This PR add argument `worker_shim_pid`, it is incompatible.

**Comment by ericl** [comment]
@iycheng could you review the approach here?

**Comment by architkulkarni** [comment]
> The CI `doc:doc_code_runtime_env_example` always failed , because it will install ray[serve] release version. This PR add argument `worker_shim_pid`, it is incompatible.

Sorry, this is my fault.  When I added the test, it was only supposed to be run in the post-wheel-build tests (indeed, it passed there https://buildkite.com/ray-project/ray-builders-pr/builds/7104#669f8e90-bd62-42fe-981b-70a5452e1cc0/6-10471) but I forgot to disable the test in buildkite/ray-builders-pr/book-doc-tests-and-examples, which is run before the wheel is built.  Let me make a quick PR to fix this.

**Comment by architkulkarni** [comment]
Here's the PR that will fix CI: https://github.com/ray-project/ray/pull/16439

**Comment by architkulkarni** [comment]
It looks like there are some other errors in the Windows build: https://github.com/ray-project/ray/pull/16398/checks?check_run_id=2829232526#step:7:880. I don't know why it's only happening for Windows though.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -901,7 +901,15 @@ void CoreWorker::RegisterToGcs() {
 }
 
 void CoreWorker::CheckForRayletFailure() {
-  if (!IsParentProcessAlive()) {
+  // When running worker process in container, the worker parent process is not raylet.
+  // So we add RAYLET_PID enviroment to ray worker process.
```
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:905]: Should we always do this? @ericl

**Comment by iycheng** [comment]
I have one question here: can we use the worker pid instead of shim id in the raylet? For example, after registering, we use the worker id instead?

I think my concern here is that there are shim process id and pid going around, and I'm wondering whether we can simplify the logic there.

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -901,7 +901,15 @@ void CoreWorker::RegisterToGcs() {
...
+  if (const char *env_pid = std::getenv("RAYLET_PID")) {
+    pid_t pid = static_cast<pid_t>(std::atoi(env_pid));
+    if (!IsProcessAlive(pid)) {
+      RAY_LOG(ERROR) << "Raylet failed. Shutting down.Raylet Pid: " << pid;
```
Comment (architkulkarni) [inline src/ray/core_worker/core_worker.cc:909]:
```suggestion
      RAY_LOG(ERROR) << "Raylet failed. Shutting down. Raylet PID: " << pid;
```

**Review by architkulkarni — COMMENTED** [review]
Thanks for this PR and for adding the test!  If I understand correctly, this PR allows forked workers with different PIDs to register with the Raylet by remembering and carrying with them the PID of their ancestor shim process, which the Raylet can then check.  This will allow for greater flexibility in runtime env setup function.  The logic looks good to me. 

I share Yi's concern that it's a little confusing that the shim process is used in some places while the worker process is used in some other places.  I guess that's because the worker pool only keeps track of the processes that it starts and it doesn't know about the forked workers.

**Comment by chenk008** [comment]
> I have one question here: can we use the worker pid instead of shim id in the raylet? For example, after registering, we use the worker id instead?
> 
> I think my concern here is that there are shim process id and pid going around, and I'm wondering whether we can simplify the logic there.

@iycheng @architkulkarni Thanks for reviewing. Currently only  `raylet::WorkerPool::State  starting_worker_processes` use shim process PID, and after the worker registered, `State` will save worker process PID in WorkerInterface and don't use shim process PID anymore.

I found the following function will use `starting_worker_processes`

1. MonitorStartingWorkerProcess: it will use shim process PID to check whether the worker registration is timeout,
1. RegisterWorker:  Worker tries to register, now use worker shim process PID to validate.
1. OnWorkerStarted:  After the worker has registered,  the worker tries to `AnnounceWorkerPort`. Now use worker shim process PID.
1. StartWorkerProcess: Try to start a new worker. It is not affected.
1. TryKillingIdleWorkers:  It uses starting_worker_processes to check whether a worker process has pending registration. Now use worker shim process PID.
1. WarnAboutSize: It will print some warning messages about pending workers. It is not affected.
1. ~WorkerPool: The WorkerPool destructor will try to kill pending workers, now it will kill the shim process. There is a corner case: if the shim process was killed, but the worker process is still alive. I don't think this is a big matter, the worker process will exit soon in `CheckForRayletFailure`

File: src/ray/raylet/worker_pool_test.cc [hunk 3]
```diff
@@ -1133,6 +1141,24 @@ TEST_F(WorkerPoolTest, PopWorkerWithRuntimeEnv) {
...
+TEST_F(WorkerPoolTest, StartWorkWithDifferentShimPid) {
+  auto task_spec = ExampleTaskSpec();
+  auto worker = worker_pool_->PopWorker(task_spec);
+  RAY_CHECK(!worker);
+  auto last_process = worker_pool_->LastStartedWorkerProcess();
+
+  // Register worker with different worker PID
+  pid_t shim_pid = last_process.GetId();
+  worker = CreateWorker(Process());
+  RAY_CHECK_OK(worker_pool_->RegisterWorker(worker, shim_pid+1000, shim_pid,
```
[the remark this answers is not in the document]
Comment (chenk008) [reply src/ray/raylet/worker_pool_test.cc:1153]: I just to want use a different PID as worker process PID to run this test. Should we use `StartProcess` to get a new PID?

**Comment by chenk008** [comment]
> It looks like there are some other errors in the Windows build: https://github.com/ray-project/ray/pull/16398/checks?check_run_id=2829232526#step:7:880. I don't know why it's only happening for Windows though.

@architkulkarni I have fix build error on Windows. Now there is some msgpack errors: Unable to allocate internal buffer , it looks like a outofmemory error.

**Comment by iycheng** [comment]
@chenk008 does this mean that we should just use shim process pid there by default and once registered, we use worker pid?

We now have two pids, 1) the pid that try to start a worker; 2) the worker pid. They can be the same or not. My feeling here is that we should just update our current structure to fit this.

File: src/ray/raylet/worker_pool_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (architkulkarni) [reply src/ray/raylet/worker_pool_test.cc]: I see, thanks.  I guess the value of the worker PID isn't important for this test so it's fine to leave it the way it is.

**Comment by chenk008** [comment]
> @chenk008 does this mean that we should just use shim process pid there by default and once registered, we use worker pid?

Yes, this is the way how this PR works.

> We now have two pids, 1) the pid that try to start a worker; 2) the worker pid. They can be the same or not. My feeling here is that we should just update our current structure to fit this.

@iycheng  Do you mean rename `starting_worker_processes` to `staring_worker_shim_processes`? I think currently only this field was affected.

File: python/ray/worker.py [hunk 4]
```diff
@@ -1144,7 +1144,8 @@ def connect(node,
             job_id=None,
             namespace=None,
             job_config=None,
-            runtime_env_hash=0):
+            runtime_env_hash=0,
+            worker_shim_pid=0):
```
Comment (jovany-wang) [inline python/ray/worker.py:1148]: minor: Add a doc comment on this.

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1083,6 +1085,10 @@ void NodeManager::ProcessRegisterClientRequestMessage(
   } else {
     // Register the new driver.
     RAY_CHECK(pid >= 0);
+    if (worker_shim_pid == 0) {
```
Comment (jovany-wang) [inline src/ray/raylet/node_manager.cc:1088]: No need to set shim to driver.

File: src/ray/raylet/worker.cc [hunk 6]
```diff
@@ -67,6 +67,13 @@ void Worker::SetProcess(Process proc) {
   proc_ = std::move(proc);
 }
 
+Process Worker::GetShimProcess() const { return shim_proc_; }
```
Comment (jovany-wang) [inline src/ray/raylet/worker.cc:70]:
```C++
Process Worker::GetShimProcess() const { 
    RAY_CHECK(this is not a driver.);
    return shim_proc_; 
}

```

File: src/ray/raylet/worker_pool.cc [hunk 7]
```diff
@@ -446,20 +447,25 @@ boost::optional<const rpc::JobConfig &> WorkerPool::GetJobConfig(
...
 
 Status WorkerPool::RegisterWorker(const std::shared_ptr<WorkerInterface> &worker,
-                                  pid_t pid,
+                                  pid_t pid, pid_t worker_shim_pid,
                                   std::function<void(Status, int)> send_reply_callback) {
   RAY_CHECK(worker);
 
   auto &state = GetStateForLanguage(worker->GetLanguage());
-  auto process = Process::FromPid(pid);
+  if (worker_shim_pid == 0) {
```
Comment (jovany-wang) [inline src/ray/raylet/worker_pool.cc:455]: RAY_CHECK(worker_shim_pid > 0);

**Comment by jovany-wang** [comment]
> @chenk008 does this mean that we should just use shim process pid there by default and once registered, we use worker pid?
> 
> We now have two pids, 1) the pid that try to start a worker; 2) the worker pid. They can be the same or not. My feeling here is that we should just update our current structure to fit this.

Shim pids are only used before worker registering. So I think it's good enough that we just rename `starting_worker_processes` to `starting_shim_processes`.

File: src/ray/raylet/node_manager.cc (hunk 5 above)
Comment (chenk008) [reply src/ray/raylet/node_manager.cc]: Add comment, delete this code

File: python/ray/worker.py (hunk 4 above)
Comment (chenk008) [reply python/ray/worker.py:1148]: Add description

File: src/ray/raylet/worker.cc (hunk 6 above)
Comment (chenk008) [reply src/ray/raylet/worker.cc]: Add check in GetShimProcess

File: src/ray/raylet/worker_pool.cc (hunk 7 above)
Comment (chenk008) [reply src/ray/raylet/worker_pool.cc]: Added check in RegisterWorker

**Comment by iycheng** [comment]
@chenk008  I just have another look at the code and the logic looks clean. 
I don't think it's necessary to rename it. We can always have shim pid and worker pid in side a worker, where the first one means the pid starting the worker and the second one is the actual worker pid. If shim pid is not set in worker, it by default equal the worker pid. Then in our code, we don't need to do special things checking shim pid set or not, we can just use that directly. 

For example,  in register worker, we have things like whether shim pid is set or not and do something special, which requires some context about it. Any anyway since we've introduced shim pid, we probably just give a better definition about this and where it's used (also for worker pid) and don't mix them (like when something not set use this).
 
Btw, is there any test for this from the python layer? You can put a simple test shim process and start the worker.

**Comment by chenk008** [comment]
@iycheng I have added a python test, please take a look.

File: src/ray/raylet/worker_pool.cc [hunk 8]
```diff
@@ -446,20 +447,25 @@ boost::optional<const rpc::JobConfig &> WorkerPool::GetJobConfig(
...
 
   auto &state = GetStateForLanguage(worker->GetLanguage());
-  auto process = Process::FromPid(pid);
-
-  if (state.starting_worker_processes.count(process) == 0) {
-    RAY_LOG(WARNING) << "Received a register request from an unknown worker "
-                     << process.GetId();
+  // When the "shim process" setup worker is not needed, the worker_shim_pid will be 0.
+  if (worker_shim_pid == 0) {
+    worker_shim_pid = pid;
```
Comment (iycheng) [inline src/ray/raylet/worker_pool.cc:457]: should it always be set?

**Comment by iycheng** [comment]
Thanks! LGTM. One comment.

File: src/ray/raylet/worker_pool.cc (hunk 8 above)
Comment (chenk008) [reply src/ray/raylet/worker_pool.cc:457]: @iycheng  The `worker_shim_pid` will be set only if enable runtime env. https://github.com/alipay/ray/blob/refact_worker_pid_validate/python/ray/workers/setup_runtime_env.py#L39

Comment (jovany-wang) [reply src/ray/raylet/worker_pool.cc:457]: It can be set when creating the register message in RayletClient.

Comment (chenk008) [reply src/ray/raylet/worker_pool.cc:457]: I think `worker_shim_pid` should not be set if we don't use shim process

Comment (chenk008) [reply src/ray/raylet/worker_pool.cc]: I have moved this logic into RayletClient

**Review by jovany-wang — APPROVED** [review]

**Comment by jovany-wang** [comment]
Looks great!

**Comment by ericl** [comment]
| //python/ray/tests:test_gcs_fault_tolerance                              FAILED in 3 out of 3 in 69.3s
  | Stats over 3 runs: max = 69.3s, min = 69.2s, avg = 69.2s, dev = 0.0s
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_gcs_fault_tolerance/test.log
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_gcs_fault_tolerance/test_attempts/attempt_1.log
  | /root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_gcs_fault_tolerance/test_attempts/attempt_2.log
  |  
 

<br class="Apple-interchange-newline">

**Review by fishbone — APPROVED** [review]
Please fix the related failure. Otherwise, feel good about this.

**Comment by chenk008** [comment]
@ericl @iycheng I found `test_gcs_fault_tolerance` is marked flaky in ci.sh. Can you help me to retry this job: `python-medium-a-j` ? I can't click the retry button.

**Review by architkulkarni — APPROVED** [review]
Looks good to me, thanks for adding the test!  I re-ran the failing test, but I just noticed there are some merge conflicts unfortunately--could you please take a look?

**Comment by chenk008** [comment]
@architkulkarni  I have merged master branch to fix conflicts. Please take a look again, thanks!

**Comment by architkulkarni** [comment]
Looks good to me :) The Travis MacOS build is known the be flaky according to https://flaky-tests.ray.io/ so I think it's fine to merge this PR.

**Comment by architkulkarni** [comment]
Actually it looks like `python/ray/tests/test_basic.py::test_release_resources_race` timed out: https://travis-ci.com/github/ray-project/ray/builds/230305052#L35027 
Let me investigate this

**Comment by architkulkarni** [comment]
https://travis-ci.com/github/ray-project/ray/builds/230305052 Started the Travis CI MacOS build again, let's see what hapens

**Pull request merged**